File: frontend/src/Card.js
import React from "react";

export default function Card({ rank, suit, displayType }) {
  function colorClass() {
    return suit === 0 || suit === 2 ? "card-red" : "card-black";
  }

  function cardAsCodePoint() {
    if (rank === -1) {
      // Unicode for backface of a card
      return 0x1f0a0;
    }

    const suitOffset = codePointSuits[suit];
    // In President "3" is the lowest, so it has rank 0,
    // so add 2 to get it's rank in normal playing cards.
    var rankOffset = (rank + 2) % 13;
    if (rankOffset > 10) {
      // There's a weird "knight" card between Jack and Queen in the Unicode charset
      rankOffset++;
    }
    return suitOffset + rankOffset;
  }

  return (
    <div className={`card ${colorClass()} ${displayType}`}>
      {String.fromCodePoint(cardAsCodePoint())}
    </div>
  );
}

const codePointSuits = [
  0x1f0c1, // Ace of Diamonds
  0x1f0d1, // Ace of Clubs
  0x1f0b1, // Ace of Hearts
  0x1f0a1, // Ace of Spades
];

File: card.py
from typing import Union

FACE_RANKS = ["3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A", "2"]
FACE_SUITS = ["DIAMONDS", "CLUBS", "HEARTS", "SPADES"]

def rank(card: int) -> int:
    return card % 13

def suit(card: int) -> int:
    return card // 13

def description(card: int) -> str:
    return f"{FACE_RANKS[rank(card)]} OF {FACE_SUITS[suit(card)]}"

def serialize(card: int) -> dict[str, Union[str, int]]:
    return {
        "value": card,
        "rank": rank(card),
        "suit": suit(card),
        "description": description(card),
    }

def is_playable(card: int, top_card: int) -> bool:
    return rank(card) >= rank(top_card)
